code/modules/client/preferences2/character_save.dm

```diff
@@ -0,0 +1,550 @@
+	if(!SSdbcore.IsConnected())
+		return
+
+	if(IS_GUEST_KEY(C.ckey))
+		return
+
+	if(from_db)
+		// We exist in the DB. Prepare for a gargantuan update query
+		var/datum/DBQuery/update_query = SSdbcore.NewQuery({"
+			UPDATE [format_table_name("characters")] SET
+				species=:species,
+				real_name=:real_name,
+				name_is_always_random=:name_is_always_random,
+				body_is_always_random=:body_is_always_random,
+				gender=:gender,
+				age=:age,
+				hair_color=:hair_color,
+				gradient_color=:gradient_color,
+				facial_hair_color=:facial_hair_color,
+				skin_tone=:skin_tone,
+				hair_style_name=:hair_style_name,
+				gradient_style=:gradient_style,
+				facial_style_name=:facial_hair_style,
+				underwear=:underwear,
+				underwear_color=:underwear_color,
+				undershirt=:undershirt,
+				socks=:socks,
+				backbag=:backbag,
+				jumpsuit_style=:jumpsuit_style,
+				uplink_loc=:uplink_loc,
+				features=:features,
+				custom_names=:custom_names,
+				helmet_style=:helmet_style,
+				preferred_ai_core_display=:preferred_ai_core_display,
+				preferred_security_department=:preferred_security_department,
+				joblessrole=:joblessrole,
+				job_preferences=:job_preferences,
+				all_quirks=:all_quirks,
+				equipped_gear=:equipped_gear
+			WHERE
+				ckey=:ckey AND slot=:slot
+		"}, list(
+			// Now for the above but in a fucking monsterous list
+			"species" = pref_species.id,
+			"real_name" = real_name,
+			"name_is_always_random" = be_random_name,
+			"body_is_always_random" = be_random_body,
+			"gender" = gender,
+			"age" = age,
+			"hair_color" = hair_color,
+			"gradient_color" = gradient_color,
+			"facial_hair_color" = facial_hair_color,
+			"skin_tone" = skin_tone,
+			"hair_style_name" = hair_style,
+			"gradient_style" = gradient_style,
+			"facial_hair_style" = facial_hair_style,
+			"underwear" = underwear,
+			"underwear_color" = underwear_color,
+			"undershirt" = undershirt,
+			"socks" = socks,
+			"backbag" = backbag,
+			"jumpsuit_style" = jumpsuit_style,
+			"uplink_loc" = uplink_spawn_loc,
+			"features" = json_encode(features),
+			"custom_names" = json_encode(custom_names),
+			"helmet_style" = helmet_style,
+			"preferred_ai_core_display" = preferred_ai_core_display,
+			"preferred_security_department" = preferred_security_department,
+			"joblessrole" = joblessrole,
+			"job_preferences" = json_encode(job_preferences),
+			"all_quirks" = json_encode(all_quirks),
+			"equipped_gear" = json_encode(equipped_gear),
+			"ckey" = C.ckey,
+			"slot" = slot_number
+		))
+
+		if(!update_query.warn_execute())
+			to_chat(usr, "<span class='boldannounce'>Failed to save your character. Please inform the server operator.</span>")
+
+		qdel(update_query)
+		return
+
+	// If we are here, we dont exist in the DB, which means its insert time.
+	// GET READY FOR SOMETHING HORRIFYING. ITS MUCH WORSE THAN BEFORE
+	var/datum/DBQuery/insert_query = SSdbcore.NewQuery({"
+		INSERT INTO [format_table_name("characters")] (
```

Comment:
You can get rid of a mass of duplicate SQL if you use REPLACE instead of INSERT. (Downside: It'll inc the autoincrement counter each time, instead of only on the INSERTs, but since nothing actually uses ID anyway, you can drop that column and promote your UNIQUE INDEX to a PRIMARY KEY.)